### PR TITLE
0x05a Broadcast Receiver fix 1340

### PR DESCRIPTION
**Pull request by zehuanli**
This PR covers issue #1340

### CODE REVIEW

File: Document/0x05a-Platform-Overview.md [hunk 1]
```diff
@@ -367,12 +367,39 @@ An example Broadcast Receiver declaration with an intent filter in a manifest:
...
+// Define an intent filter with actions that the broadcast receiver listens for
+IntentFilter intentFilter = new IntentFilter();
+intentFilter.addAction("com.owasp.myapplication.MY_ACTION");
+// To register the broadcast receiver
+registerReceiver(myReceiver, intentFilter);
+// To un-register the broadcast receiver
+unregisterReceiver(myReceiver);
+```
+
+Note that Android apps don’t need to be running to receive intents; the system starts apps automatically when a relevant intent is raised.
```
Comment (commjoen) [inline Document/0x05a-Platform-Overview.md]:
```suggestion
Note that the system starts an app with the registered receiver automatically when a relevant intent is raised.
```

File: Document/0x05a-Platform-Overview.md [hunk 2]
```diff
@@ -367,12 +367,39 @@ An example Broadcast Receiver declaration with an intent filter in a manifest:
...
 After receiving an implicit intent, Android will list all apps that have registered a given action in their filters. If more than one app has registered for the same action, Android will prompt the user to select from the list of available apps.
 
 An interesting feature of Broadcast Receivers is that they are assigned a priority; this way, an intent will be delivered to all authorized receivers according to their priority.
 
 A Local Broadcast Manager can be used to make sure intents are received from the internal app only, and any intent from any other app will be discarded. This is very useful for improving security.
 
+###### Implicit Broadcast Receiver Limitiation
+
+According to [Background Execution Limits](https://developer.android.com/about/versions/oreo/background.html#broadcasts), apps that target Android 8.0 (API level 26) or higher can no longer register Broadcast Receivers for implicit broadcasts in their manifest, except for those listed in [Implicit Broadcast Exceptions](https://developer.android.com/guide/components/broadcast-exceptions). The Broadcast Receivers created at runtime by calling `Context.registerReceiver()` are not affected by this limitation.
+
```
Comment (commjoen) [inline Document/0x05a-Platform-Overview.md]: Maybe add the earlier limitation on android 7 as well? https://developer.android.com/about/versions/nougat/android-7.0-changes.html?source=post_page---------------------------#bg-opt

File: Document/0x05a-Platform-Overview.md [hunk 3]
```diff
@@ -355,7 +355,7 @@ To improve security and privacy, a Local Broadcast Manager is used to send and r
 
 Broadcast Receivers are components that allow apps to receive notifications from other apps and from the system itself. With it, apps can react to events (internal, initiated by other apps, or initiated by the operating system). They are generally used to update user interfaces, start services, update content, and create user notifications.
 
-Broadcast Receivers must be declared in the Android Manifest file. The manifest must specify an association between the Broadcast Receiver and an intent filter to indicate the actions the receiver is meant to listen for. If Broadcast Receivers aren't declared, the app won't listen to broadcasted messages. However, apps don’t need to be running to receive intents; the system starts apps automatically when a relevant intent is raised.
+There are two ways to make a Broadcast Receiver known to the system. One way is to declare it in the Android Manifest file. The manifest should specify an association between the Broadcast Receiver and an intent filter to indicate the actions the receiver is meant to listen for.
```
Comment (commjoen) [inline Document/0x05a-Platform-Overview.md]: Can you explain the difference between explicit and implicit broadcast receivers here please?

File: Document/0x05a-Platform-Overview.md [hunk 4]
```diff
@@ -367,12 +367,39 @@ An example Broadcast Receiver declaration with an intent filter in a manifest:
...
+IntentFilter intentFilter = new IntentFilter();
+intentFilter.addAction("com.owasp.myapplication.MY_ACTION");
+// To register the broadcast receiver
+registerReceiver(myReceiver, intentFilter);
+// To un-register the broadcast receiver
+unregisterReceiver(myReceiver);
+```
+
+Note that Android apps don’t need to be running to receive intents; the system starts apps automatically when a relevant intent is raised.
+
```
Comment (commjoen) [inline Document/0x05a-Platform-Overview.md]: Another thing: what does the programmatically declared receiver mean in terms of android:exported and android:enabled? How can we make sure that not everything can start this?

File: Document/0x05a-Platform-Overview.md (hunk 2 above)
Comment (commjoen) [reply Document/0x05a-Platform-Overview.md]: And limitations for android 9 as well: https://developer.android.com/guide/components/broadcasts

File: Document/0x05a-Platform-Overview.md [hunk 5]
```diff
@@ -367,12 +367,39 @@ An example Broadcast Receiver declaration with an intent filter in a manifest:
 </receiver>
 ```
 
+The other way is to create the receiver dynamically in code and register it with the `Context.registerReceiver()` method.
+
```
Comment (commjoen) [inline Document/0x05a-Platform-Overview.md]: I know it is a little bit out of the way of your PR, but can I ask you a favor? Can you link to https://github.com/OWASP/owasp-mstg/blob/master/Document/0x05h-Testing-Platform-Interaction.md#broadcast-permission-enforcement to show how security matters and update that section with the security considerations of https://developer.android.com/guide/components/broadcasts#security-and-best-practices ?

File: Document/0x05a-Platform-Overview.md [hunk 6]
```diff
@@ -367,12 +367,39 @@ An example Broadcast Receiver declaration with an intent filter in a manifest:
 </receiver>
 ```
 
+The other way is to create the receiver dynamically in code and register it with the `Context.registerReceiver()` method.
+
+An example of registering Broadcast Receiver dynamically:
```
Comment (TheDauntless) [inline Document/0x05a-Platform-Overview.md]:
```suggestion
An example of registering a Broadcast Receiver dynamically:
```

File: Document/0x05a-Platform-Overview.md [hunk 7]
```diff
@@ -367,12 +367,39 @@ An example Broadcast Receiver declaration with an intent filter in a manifest:
...
+
 After receiving an implicit intent, Android will list all apps that have registered a given action in their filters. If more than one app has registered for the same action, Android will prompt the user to select from the list of available apps.
 
 An interesting feature of Broadcast Receivers is that they are assigned a priority; this way, an intent will be delivered to all authorized receivers according to their priority.
 
 A Local Broadcast Manager can be used to make sure intents are received from the internal app only, and any intent from any other app will be discarded. This is very useful for improving security.
 
+###### Implicit Broadcast Receiver Limitiation
+
+According to [Background Execution Limits](https://developer.android.com/about/versions/oreo/background.html#broadcasts), apps that target Android 8.0 (API level 26) or higher can no longer register Broadcast Receivers for implicit broadcasts in their manifest, except for those listed in [Implicit Broadcast Exceptions](https://developer.android.com/guide/components/broadcast-exceptions). The Broadcast Receivers created at runtime by calling `Context.registerReceiver()` are not affected by this limitation.
```
Comment (TheDauntless) [inline Document/0x05a-Platform-Overview.md]:
```suggestion
According to [Background Execution Limits](https://developer.android.com/about/versions/oreo/background.html#broadcasts "Background Execution Limits"), apps that target Android 8.0 (API level 26) or higher can no longer register Broadcast Receivers for implicit broadcasts in their manifest, except for those listed in [Implicit Broadcast Exceptions](https://developer.android.com/guide/components/broadcast-exceptions "Implicit Broadcast Exceptions"). The Broadcast Receivers created at runtime by calling `Context.registerReceiver()` are not affected by this limitation.
```

File: Document/0x05a-Platform-Overview.md (hunk 3 above)
Comment (cpholguera) [reply Document/0x05a-Platform-Overview.md]: also consider referencing https://developer.android.com/guide/components/broadcasts.html#security-and-best-practices

File: Document/0x05a-Platform-Overview.md (hunk 4 above)
Comment (cpholguera) [reply Document/0x05a-Platform-Overview.md]: I'd update both 
- https://github.com/OWASP/owasp-mstg/blob/master/Document/0x05h-Testing-Platform-Interaction.md#inspect-the-androidmanifest-2: see above comment on the default values when setting a manifest-based receiver.
- https://github.com/OWASP/owasp-mstg/blob/master/Document/0x05h-Testing-Platform-Interaction.md#inspect-the-source-code-2: add what is the default and that this has to be considered when reviewing the code.

@commjoen Do we need an extra ticket for this or better include it in this PR?

**Comment by cpholguera** [comment]
should we include here some info about sticky broadcasts? deprecated in API level 21:

```
Sticky broadcasts should not be used. They provide no security (anyone can access them), no protection (anyone can modify them), and many other problems. The recommended pattern is to use a non-sticky broadcast to report that something has changed, with another mechanism for apps to retrieve the current value whenever desired.
```

https://developer.android.com/reference/android/content/Context.html#sendStickyBroadcast(android.content.Intent)

**Comment by commjoen** [comment]
https://github.com/OWASP/owasp-mstg/pull/1355#discussion_r310362148 : @zehuanli : do you want to pick it up here? or should we create a new ticket :) ?

File: Document/0x05a-Platform-Overview.md (hunk 2 above)
Comment (zehuanli) [reply Document/0x05a-Platform-Overview.md]: Description for both added.

File: Document/0x05a-Platform-Overview.md (hunk 3 above)
Comment (zehuanli) [reply Document/0x05a-Platform-Overview.md]: The only reference I could find (and added) about "implicit broadcast" in Android Dev is under [Manifest-declared receivers](https://developer.android.com/guide/components/broadcasts.html#manifest-declared-receivers):

> implicit broadcasts (broadcasts that do not target your app specifically)

My understanding for "implicit broadcast receiver" is that it doesn't only listen for broadcasts that target the app itself, but also for those "implicit broadcasts". However, there's no reference I could find in the documentation.

Comment (zehuanli) [reply Document/0x05a-Platform-Overview.md]: @cpholguera Is there a particular point you think should be imported from the "Security considerations and best practices" paragraph, or are you suggesting adding the whole page as a reference?

Comment (commjoen) [reply Document/0x05a-Platform-Overview.md]: @zehuanli these are implicit broadcasts: https://developer.android.com/guide/components/broadcast-exceptions, basically most of them are system events where an app can listen to and act upon.

Comment (cpholguera) [reply Document/0x05a-Platform-Overview.md]: At least referencing to that page as a link should be sufficient, thank you :)

File: Document/0x05a-Platform-Overview.md (hunk 4 above)
Comment (commjoen) [reply Document/0x05a-Platform-Overview.md]: What would you prefer @zehuanli ? would you be able to update these parts? or would you rather have it in a separate isue?

Comment (zehuanli) [reply Document/0x05a-Platform-Overview.md]: @commjoen Yes I think I can update above parts here as they're kinda related to this issue.

Comment (commjoen) [reply Document/0x05a-Platform-Overview.md]: Thank you for taking care of it @zehuanli 👍

File: Document/0x05a-Platform-Overview.md (hunk 3 above)
Comment (commjoen) [reply Document/0x05a-Platform-Overview.md]: Thanks!

Comment (zehuanli) [reply Document/0x05a-Platform-Overview.md]: Link to  [Security considerations and best practices](https://developer.android.com/guide/components/broadcasts.html#security-and-best-practices) has been added.

@commjoen description of "implicit broadcast" has been added in [commit](https://github.com/zehuanli/owasp-mstg/commit/72694f066d156e02cce65f1fef837a9fad605564). Changes of implicit broadcasts in [Broadcast Exceptions](https://developer.android.com/guide/components/broadcast-exceptions) has been added in [commit](https://github.com/zehuanli/owasp-mstg/commit/441e0a699dbd61d0696fb9d618af99f23f0e25ca). Please let me know if there are any changes needed.

**Comment by zehuanli** [comment]
Link to  [Security considerations and best practices](https://developer.android.com/guide/components/broadcasts.html#security-and-best-practices) has been added.

@commjoen description of "implicit broadcast" has been added in [commit](https://github.com/zehuanli/owasp-mstg/commit/72694f066d156e02cce65f1fef837a9fad605564). Changes of implicit broadcasts in [Broadcast Exceptions](https://developer.android.com/guide/components/broadcast-exceptions) has been added in [commit](https://github.com/zehuanli/owasp-mstg/commit/441e0a699dbd61d0696fb9d618af99f23f0e25ca). Please let me know if there are any changes needed.

**Comment by zehuanli** [comment]
No problem! Please don't merge now as there are still some pending changes. I'm still sorting them out and will update soon.

**Comment by zehuanli** [comment]
TODO:

1. What does the programmatically declared receiver mean in terms of android:exported and android:enabled? How can we make sure that not everything can start this?
 - https://github.com/OWASP/owasp-mstg/blob/master/Document/0x05h-Testing-Platform-Interaction.md#inspect-the-androidmanifest-2: update this regarding default values when setting a manifest-based receiver
 - https://github.com/OWASP/owasp-mstg/blob/master/Document/0x05h-Testing-Platform-Interaction.md#inspect-the-source-code-2: add what is the default and that this has to be considered when reviewing the code

2. Link to https://github.com/OWASP/owasp-mstg/blob/master/Document/0x05h-Testing-Platform-Interaction.md#broadcast-permission-enforcement to show how security matters and update that section with the security considerations of https://developer.android.com/guide/components/broadcasts#security-and-best-practices

**Comment by commjoen** [comment]
Thanks! Let us know when you are ready! 👍

**Comment by commjoen** [comment]
Let's create a new ticket for your open todo's: https://github.com/OWASP/owasp-mstg/issues/1495 and assign it to you. In the meantime @TheDauntless can you please approve?

**Pull request merged**